File: 2115-find-all-possible-recipes-from-given-supplies/2115-find-all-possible-recipes-from-given-supplies.java
class Solution {
    public List<String> findAllRecipes(String[] rec, List<List<String>> ing, String[] sup) {
        int n = rec.length;
        HashSet<String> st = new HashSet<>(Arrays.asList(sup));
        HashMap<String , Integer> mp = new HashMap<>();
        HashMap<String , List<Integer>> adj = new HashMap<>();
        int[] ind = new int[n];
        Queue<Integer> q = new LinkedList<>();
        for(int i = 0 ; i < n; i++){
            for(String s : ing.get(i)){
                if(!st.contains(s)){
                    adj.putIfAbsent(s , new ArrayList<>());
                    adj.get(s).add(i);
                    ind[i]++;
                }
            }
        }
        for(int i = 0 ; i < n ; i++){
            if(ind[i] == 0) q.offer(i);
        }
        List<String> ans = new ArrayList<>();
        while(!q.isEmpty()){
            int i = q.peek();
            q.poll();
            ans.add(rec[i]);
            for(int j : adj.getOrDefault(rec[i] , new ArrayList<>())){
                ind[j]--;
                if(ind[j] == 0) q.offer(j);
            }
        }
        return ans;
    }
}

File: 2115-find-all-possible-recipes-from-given-supplies/2115-find-all-possible-recipes-from-given-supplies.cpp
class Solution {
public:
    vector<string> findAllRecipes(vector<string>& rec, vector<vector<string>>& ing, vector<string>& sup) {
        unordered_set<string> st(sup.begin() , sup.end());
        unordered_map<string , vector<int>> adj;
        int n = rec.size();
        vector<int> ind(n , 0);
        for(int i = 0 ; i < n ; i++){
            for(auto& s : ing[i]){
                if(!st.count(s)){
                    adj[s].push_back(i);
                    ind[i]++;
                }
            }
        }

        queue<int> q;
        for(int i = 0 ;i < n ; i++){
            if(ind[i] == 0) q.push(i);
        }
        vector<string> ans;
        while(!q.empty()){
            int node = q.front();
            q.pop();
            ans.push_back(rec[node]);
            for(auto& e : adj[rec[node]]){
                ind[e]--;
                if(ind[e] == 0)q.push(e);
            }
        }
        return ans;
    }
};